---
title: "DataPreprocessing"
author: "Ye Bi"
date: "`r Sys.Date()`"
output: html_document
---

- preprocessing for metabolites and SNPs dataset.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "~/Library/CloudStorage/OneDrive-VirginiaTech/Research/Codes/research/RiceUNLMetabolites/GBLUP4Met/Githubs/GBLUP4Met")

```

# loading packages
```{r, message=F}
library(tidyverse)
library(lme4)
library(readxl)
library(devtools)
library(ggbiplot)
library(pheatmap)
library(tidyverse)
library(ggplot2)
library(ggpubr)
library(ggfortify)#draw pca plots
library(data.table)

met_path = "../../Met"
geno_path = "../../Geno"
```

# 1. Metabolite data cleaning.
- Keep "NSFTV.Id"  "Treatment" "Ext.batch" "Run" and 73 Metabolites
- Remove NA, duplicated, no matched genotypes (NSFTV_ID). 
```{r, eval=F}
# met22 <- read_excel("../Met/raw_data/221019_metabolite raw data.xlsx")
# met22_info <- read_excel("../Met/raw_data/221019_metabolite raw data.xlsx", 
#                          sheet = "Sample information")

## txt file is easier to implement.
met22 <- read.delim("../Met/raw_data/221019_metabolite_raw_data.txt")

met <- met22[,c(7,2,4,5,12:84)]
colnames(met)[1:4] = c("NSFTV_ID",  "Treatment", "Batch", "Run")
      # "NSFTV_ID"  "Treatment" "Batch" "Run" and Metabolites

met = met[!is.na(met$NSFTV_ID),] # one NA in NSFTV_ID
met$NSFTV_ID= paste0("NSFTV_", met$NSFTV_ID) 
met <- met %>% arrange(NSFTV_ID)

## Change met names into a1-h3.
name = c(paste0("a",1:10),paste0("b",1:10),paste0("c",1:10),paste0("d",1:10),
         paste0("e",1:10),paste0("f",1:10),paste0("g",1:10),paste0("h",1:3))
colnames(met)[5:77] = name
metL <- met %>% group_split(Treatment);names(metL) <- c("Control", "Stress")

#remove duplicated genotypes in met.stress 
#no dupicated genotypes in control group
sum(duplicated(metL[[1]]$NSFTV_ID))
sum(duplicated(metL[[2]]$NSFTV_ID)) #duplicate in stress group
dd = duplicated(metL[[2]]$NSFTV_ID)
metL[[2]] = metL[[2]][!dd, ]


id_common = intersect(metL[[1]]$NSFTV_ID, metL[[2]]$NSFTV_ID)
metL[[1]] = metL[[1]] %>% filter(NSFTV_ID %in% id_common) %>% droplevels()
metL[[2]] = metL[[2]] %>% filter(NSFTV_ID %in% id_common) %>% droplevels()
all(metL[[1]]$NSFTV_ID == metL[[2]]$NSFTV_ID)
```

# 2. Match genotype, met datasets.
```{r, eval = F}
######################################################
## Match genotyped, metabolited data set.
######################################################

## Read geno data.
geno_700K <- readRDS("../Geno/raw_data/geno_final_700K.rds")
map <- readRDS("../Geno/raw_data/mapinfo_final_700k.rds")
SNP_names <- map$V2
colnames(geno_700K) = SNP_names
geno_id <- rownames(geno_700K)

#function to find out overlap with genotype data.
NSFTV_ID_met_geno_match_func <- function(MetList){
  NSFTV_IDs_match <- MetList$NSFTV_ID[MetList$NSFTV_ID %in% geno_id]
  return(NSFTV_IDs_match)
  }

met_control0 <- metL[[1]] %>% arrange(NSFTV_ID) %>% filter(NSFTV_ID %in% NSFTV_ID_met_geno_match_func(metL[[1]]))  
met_stress0 <- metL[[2]] %>% arrange(NSFTV_ID) %>% filter(NSFTV_ID %in% NSFTV_ID_met_geno_match_func(metL[[2]]))


met_NA_func <- function(met, met2remove){
  
            ## Impute NA in metabolite data with colmean
            col_means = colMeans(met[,-c(1:4)], na.rm = T)
            for(i in colnames(met)[-c(1:4)]){
              aa = table(is.na(met[,i]))
              bb=which(is.na(met[,i]))
              met[bb,i] = col_means[i]
            }
            sum(is.na(met[,5:77]))
            
            met_net = met[,-c(1:4)]
            met = cbind(met[,1:4], met_net[,-met2remove]) #remove several metabolites
            return(met)
            }

met_control = met_NA_func(met=met_control0, met2remove=c(13,28,30,33,38,62,67))
met_stress = met_NA_func(met=met_stress0, met2remove=c(13,28,30,33,38,62,67))
```


# 3. Adjust metabolites
```{r, eval = F}
# control group: Set Run, Batch, NSFTV_ID as factors.
met_control <- met_control %>% mutate(across(c(Run, Batch, NSFTV_ID), as.factor))

# stress group: Set Run, Batch, NSFTV_ID as factors.
met_stress <- met_stress %>% mutate(across(c(Run, Batch, NSFTV_ID), as.factor))

correct_met_func <- function(x0, n_met = 66, treatment, tf = TRUE, prepo){
  
  #log transformation  
    if(tf == TRUE){
      cat("Did log-transform for dataset \n")
      x = cbind(x0[,1:4], log(x0[,-c(1:4)]+1e-3))
    
    } else {
      x = x0
    }
  
  #Remove systematic effects
    rr = matrix(NA, nrow = nrow(x), ncol = n_met); colnames(rr) = colnames(x)[-c(1:4)]
    for (i in colnames(rr)){
      # i="a1"
      cat("Now is running ", i, "\n")
      
      if(prepo == "BLUE"){
        formu = as.formula(paste0(i, " ~ 1+(1|Run) + (1|Batch)"))
        fit1 = lmer(formula = formu, data = x)
        # print(VarCorr(fit1))
        rrr = residuals(fit1)
        rr[,i] = rrr
      }
      
      # if(prepo == 'rBLUP'){ 
      #   tryCatch({
      #     fit2 = mmer(fixed = as.formula(paste0(i, "~1")),
      #                 random = ~ Run + Batch + NSFTV_ID, data = x, verbose = F, tolParConvLL = 10)
      #   }, error=function(e){
      #     fit2 <<- list()
      #     cat("ERROR :",conditionMessage(e), "\n")
      #   })
      #   
      #   PEV = diag(fit2$PevU$NSFTV_ID[[i]])
      #   varG = fit2$sigma$NSFTV_ID
      #   BLUPs = fit2$U$NSFTV_ID[[i]]
      #   ds_met = BLUPs / (1 - PEV/as.numeric(varG)) #dblup
      #   rrr = sign(ds_met) * abs(ds_met)^(1/3) #cube root
      #   if(length(rrr)==0){
      #     rr[,i] = NA
      #   }else(rr[,i] = rrr)
      # }
      
    }
    
    met.rr = cbind.data.frame(x[,1:2], rr)
    write.csv(met.rr, paste0("../Met/met_rr_",treatment, ".csv"), row.names = F, quote = F)
    # write.csv(met.rr, paste0("../temp/lmer/met_rr_",treatment, ".csv"), row.names = F, quote = F)
    return (met.rr)
}

met_rr_control = correct_met_func(x0 = met_control, n_met = 66, treatment="Control", tf = T, prepo = "BLUE")
met_rr_stress = correct_met_func(x0 = met_stress, n_met = 66, treatment="Stress", tf = T, prepo = 'BLUE')
```

###Save met files with original names.
```{r}
met_rr_control <- read.csv("../Met/met_rr_control.csv")
met_rr_stress <- read.csv("../Met/met_rr_stress.csv")

source("./functions.R")
met_names = met_name_func(mode = "name")
colnames(met_rr_control)[-c(1:2)] = met_names
colnames(met_rr_stress)[-c(1:2)] = met_names

write.csv(met_rr_control, "../Met/met_rr_control_named.csv", row.names = F, quote = F)
write.csv(met_rr_stress, "../Met/met_rr_stress_named.csv", row.names = F, quote = F)
```

## Crossvalidation design -- randon repeat subsampling
```{r}
# 1. loading data
## 1.1 Read met data
met_rr_control <- read.csv("../Met/met_rr_control.csv")
met_rr_stress <- read.csv("../Met/met_rr_stress.csv")

# set cross-validation parameters
ntst  <-  32 # testing 162*0.2
ntrn  <-  162-32 # training 
ntrn_trt <- 162-32
nCV <- 100 # times of CV

# Cross-Validation
for (i in 1:nCV) {
  # random-sampling to decide testing & reference accessions
  cat("Now running nCV = ", i, "\n")
  #i = 1
  con = met_rr_control
  trt = met_rr_stress
  set.seed(100 + i)
  test.ix.con <- sample(1:nrow(con), size=ntst) # random sampling ntst
  train.ix.con = setdiff(1:nrow(con), test.ix.con) 
  test.ix.trt <-  sample(1:nrow(trt), size=ntst) 
  train.ix.trt = setdiff(1:nrow(trt), test.ix.trt) 
  test = rbind.data.frame(con[test.ix.con,], trt[test.ix.trt,])
  train = rbind.data.frame(con[train.ix.con,], trt[train.ix.trt,])

  train$set = "train"
  test$set = "test"
  table(test$Treatment); table(train$Treatment) 
  cv = rbind.data.frame(train, test)
  cv = dplyr::arrange(cv, Treatment, NSFTV_ID, by_group = T)
  
  dir.create(paste0("../Met/CrossValidation/cv_",i))
  write.csv(cv, paste0("../Met/CrossValidation/cv_",i,"/met_cv_",i,".csv"), row.names = F, quote = F)
}

```

# 4. Genotype data cleaning.
```{r, eval = F}
## Read into met.rr dataset to get NSFTV_ID in treatments.
met_rr_control <- read.csv("../Met/met_rr_control.csv")
met_rr_stress <- read.csv("../Met/met_rr_stress.csv")
id_control <- met_rr_control$NSFTV_ID
id_stress <- met_rr_stress$NSFTV_ID
id_comb <- union(id_control, id_stress)

## Read into raw genotype data.
geno_700k <- readRDS("../Geno/raw_data/geno_final_700K.rds") #SNPs
map <- readRDS("../Geno/raw_data/mapinfo_final_700k.rds") #map
map <- map[,c(2,1,4)]
colnames(map) <- c("SNPName","Chromosome","Position")
colnames(geno_700k) = map$SNPName

## QC for geno
geno <- geno_700k[id_comb,]
#MAF check
p <- colMeans(geno)/2
maf <- ifelse(p> 0.5, 1-p, p)
maf.index <- which(maf < 0.05) 
geno_rr <- geno[,-maf.index] # 162 386989
map_rr <- map[-maf.index, ] # 386989      3
all(map_rr$SNPName == colnames(geno_rr))
save(geno_rr, file="../Geno/geno.rr.RData")
save(map_rr, file = "../Geno/map.rr.RData")


## Convert Geno into G matrix.
load("../Geno/geno.rr.RData")
geno <- geno_rr
geno_control = geno[id_control, ] 
geno_stress = geno[id_stress, ] 
Xcs_control = scale(geno_control, center = T, scale = T)
Xcs_stress = scale(geno_stress, center = T, scale = T)
G_control = tcrossprod(Xcs_control)/ncol(Xcs_control)
G_stress = tcrossprod(Xcs_stress)/ncol(Xcs_stress)
GL = list(Control = G_control,
          Stress = G_stress)
save(GL, file='../Geno/GL.RData')

## Also G matrix for overlaped genotypes
id_op <- intersect(id_control, id_stress)
geno_op <- geno[id_op, ] # 162
Xcs_op <- scale(geno_op, center = T, scale = T)
G_op = tcrossprod(Xcs_op)/ncol(Xcs_op)
save(G_op, file = '../Geno/Gop.RData')
```

##Save G.rr in csv formate for python using.
```{r}
load("../Geno/GL.RData")
G.rr = GL$Control
write.csv(G.rr, "../Geno/G.rr.csv", quote = F, row.names = T)

write.csv(geno.rr, "../Geno/geno.rr.csv", quote = F, row.names = T)
```

# #############################################
## Draw PCA plot for subpopulation
# #############################################
## Loading files
```{r}
met_rr_control <- read.csv("../../Met/met_rr_control_named.csv")
met_rr_stress <- read.csv("../../Met/met_rr_stress_named.csv")
id_control <- met_rr_control$NSFTV_ID
rownames(met_rr_control) = met_rr_control$NSFTV_ID
rownames(met_rr_stress) = met_rr_stress$NSFTV_ID

Mcs.con = scale(met_rr_control[,-c(1:2)]) 
Mchnt.con = tcrossprod(Mcs.con)/ncol(Mcs.con)
M.rr.con = cbind.data.frame(NSFTV_ID = rownames(Mchnt.con), Mchnt.con)

Mcs.trt = scale(met_rr_stress[,-c(1:2)]) 
Mchnt.trt = tcrossprod(Mcs.trt)/ncol(Mcs.trt)
M.rr.trt = cbind.data.frame(NSFTV_ID = rownames(Mchnt.trt), Mchnt.trt)

## Convert Geno into G matrix.
load("../../Geno/geno.rr.RData") #geno_rr
geno <- geno_rr
# map <- map_rr
load("../../Geno/GL.RData")
G.rr = GL$Control
G.rr = cbind.data.frame(NSFTV_ID = rownames(G.rr), G.rr)
```
## sub info
```{r}
#read into subpop info
panel_info_clean <- read.delim("../../Geno/raw_data/panel_info_clean.txt")
subpop <- panel_info_clean %>% filter(Sample.set == "RDP1")
subpop$Other.accession.ID = gsub("NSFTV", "NSFTV_", subpop$Other.accession.ID)
subpop$fastStructure.subpopulation.call = tools::toTitleCase(subpop$fastStructure.subpopulation.call)

table(subpop$Other.accession.ID %in% rownames(G.rr))
table(rownames(G.rr) %in% subpop$Other.accession.ID) #all in suspop txt file.

table(subpop$Other.accession.ID %in% M.rr.con$NSFTV_ID)
table(M.rr.con$NSFTV_ID %in% subpop$Other.accession.ID) #all in suspop txt file.
all(M.rr.con$NSFTV_ID == M.rr.trt$NSFTV_ID)
table(met_rr_control$NSFTV_ID %in% subpop$Other.accession.ID)
```

## Function for combine geno or met and subpop in one dataframe.
```{r}
pca_sub_func=function(MM, subpop){
            pop.df = subpop[subpop$Other.accession.ID %in% MM$NSFTV_ID,]
            pop.df = data.frame(NSFTV_ID = pop.df$Other.accession.ID,
                                  Subpopu = pop.df$fastStructure.subpopulation.call)

            # MM.df = data.frame(NSFTV_ID = MM$NSFTV_ID, MM)
            MM.suspop = merge(MM, pop.df)
            return(MM.suspop)
            }
```

```{r}
G.suspop = pca_sub_func(MM=G.rr, subpop=subpop)
table(G.suspop$Subpopu)

met_control.suspop = pca_sub_func(MM=M.rr.con, subpop=subpop)
table(met_control.suspop$Subpopu)

met_stress.suspop = pca_sub_func(MM=M.rr.trt, subpop=subpop)
table(met_stress.suspop$Subpopu)


suspop.df = data.frame(NSFTV_ID = met_control.suspop$NSFTV_ID,  
                       Subpopu = met_control.suspop$Subpopu)
# write.csv(suspop.df, file = file.path("../../../Met/raw_data/subpopulation.csv"), row.names = F, quote = F)
```

```{r, echo = TRUE, eval=TRUE}
pca_plot_drawing_func = function(suspop, name){
  
  suspop = suspop %>% group_by(Subpopu) %>% mutate(Subpopu = paste0(Subpopu, ":", n()))
  
  cc.pca = prcomp(suspop[,!(colnames(suspop) %in% c('NSFTV_ID', 'Subpopu'))], scale. = T)
  
  p = autoplot(cc.pca, data = suspop, colour = "Subpopu", size = 2.5) + 
    theme_bw()+
    labs(color = "Subpopulation")+
    scale_color_brewer(palette = "Dark2")
  
  p
  return(p)
}

p_G = pca_plot_drawing_func(suspop = G.suspop, name = 'geno')
p_G
dev.print(pdf, file="../../temp/pca_geno_comb.pdf", height=6, width=9)
```


```{r, echo = TRUE, eval=TRUE}
p_met_control = pca_plot_drawing_func(suspop = met_control.suspop, name = 'met_control')
p_met_control
p_met_stress = pca_plot_drawing_func(suspop = met_stress.suspop, name = 'met_stress')
p_met_stress

library(ggpubr)
ggarrange(NULL,
          p_met_control, 
          NULL,
          p_met_stress, 
          nrow = 4, 
          heights = c(0.2,3,0.2,3),
          labels = c("","(A)","", "(B)"), vjust = c(0,-0.5,0,0), #To show the entire (A) label, adjuest vjust and add NULL
          # vjust=-0.1,
          legend = "right",
          common.legend = T)
dev.print(pdf, file="../../temp/pca_met.pdf", height=10, width=7)

```

## Boxplot for metabolite within subpopulation groups
```{r}
met_control.suspop = pca_sub_func(MM=met_rr_control, subpop=subpop)
table(met_control.suspop$Subpopu)

met_stress.suspop = pca_sub_func(MM=met_rr_stress, subpop=subpop)
table(met_stress.suspop$Subpopu)

met_control.suspop.long= pivot_longer(met_control.suspop, names_to = "Met", values_to = 'Met_value', cols = -c("NSFTV_ID", "Treatment", "Subpopu"))
met_control.suspop.long$Subpopu = as.factor(met_control.suspop.long$Subpopu)

met_stress.suspop.long= pivot_longer(met_stress.suspop, names_to = "Met", values_to = 'Met_value', cols = -c("NSFTV_ID", "Treatment", "Subpopu"))
met_stress.suspop.long$Subpopu = as.factor(met_stress.suspop.long$Subpopu)

met.suspop.long = rbind(met_control.suspop.long, met_stress.suspop.long)

met.suspop.long = met.suspop.long %>% filter(Subpopu %in% tools::toTitleCase(c('aus', 'indica', 'temperate-japonica', 'tropical-japonica')))
library(ggplot2)
met.suspop.long$Subpopu = as.factor(met.suspop.long$Subpopu)
met.suspop.long$Treatment = factor(met.suspop.long$Treatment, levels=c('Stress', 'Control'))
ggplot(met.suspop.long, aes(x = Subpopu, y = Met_value, fill = Treatment)) +
  geom_boxplot() +
  labs(x = "Subpopulation", y = "Metabolite accumulation") +
  theme_minimal()

# Perform ANOVA

anova_result <- aov(Met_value ~ Subpopu, data = met.suspop.long)

# View the summary of the ANOVA
summary(anova_result)

# Perform Tukey's HSD test
tukey_result <- TukeyHSD(anova_result)

# View the Tukey HSD results
print(tukey_result)

```

## Cross-Validation design -- leave-one-subpopulation-out
```{r}
# 1. loading data
## 1.1 Read met data
met_rr_control <- read.csv(file = file.path(met_path, '/met_rr_control_named.csv'))
met_rr_stress <- read.csv(file = file.path(met_path, '/met_rr_stress_named.csv'))
subpopulation <- read.csv(file = file.path(met_path, "/raw_data/subpopulation.csv"))
subss = c('aus', 'indica', 'temperate-japonica', 'tropical-japonica')

nCV <- as.numeric(length(subss)) # times of CV

# Cross-Validation
for (i in 1:nCV) {
  # random-sampling to decide testing & reference accessions
  cat("Now running nCV: ", subss[i], "\n")
  #i = 1
  con = met_rr_control
  trt = met_rr_stress
  
  test.ix.con <- which(subpopulation$Subpopu %in% subss[i])
  train.ix.con = setdiff(1:nrow(con), test.ix.con) 
  test.ix.trt <- which(subpopulation$Subpopu %in% subss[i])
  train.ix.trt = setdiff(1:nrow(trt), test.ix.trt) 
  
  
  test = rbind.data.frame(con[test.ix.con,], trt[test.ix.trt,])
  train = rbind.data.frame(con[train.ix.con,], trt[train.ix.trt,])

  train$set = "train"
  test$set = "test"
  table(test$Treatment); table(train$Treatment) 
  cv = rbind.data.frame(train, test)
  cv = dplyr::arrange(cv, Treatment, NSFTV_ID, by_group = T)
  cv = merge(cv, subpopulation, by = 'NSFTV_ID')
  
  cv_dir_path = file.path(paste0(met_path, "/CrossValidation_subpop/cv_",i))
  dir.create(cv_dir_path)
  write.csv(cv, paste0(cv_dir_path, "/met_cv_",i,".csv"), row.names = F, quote = F)
}

```



### splite SNP matrix by chromosome
```{r, eval=FALSE}
met_rr_control <- read.csv("../Met/met_rr_control.csv")
met_rr_stress <- read.csv("../Met/met_rr_stress.csv")
id_control <- met_rr_control$NSFTV_ID
id_stress <- met_rr_stress$NSFTV_ID
id_comb <- union(id_control, id_stress)


## Convert Geno into G matrix.
load("../Geno/geno.rr.RData") #geno_rr
load('../Geno/map.rr.RData') #map_rr
geno <- geno_rr
map <- map_rr

geno_control = geno[id_control, ] 
geno_stress = geno[id_stress, ] 

genoL = list(Control = geno_control,
             Stress = geno_stress)

# SNPchrLL <- vector('list', 2); names(SNPchrLL) = c("Control", "Stress")
SNPchrLL <- list()

# for (j in c('Control')){
  j = "Control"
  SNPchrL <- vector('list',12); names(SNPchrL) = paste0("Chr", 1:12)
  for (i in 1:12){
      geno = genoL[[j]]    
      SNP_chr = map[which(map$Chromosome == i), 'SNPName']
      # SNP_chr_others = map[which(map$Chromosome != i), 'SNPName']
      cat('number of SNP in Chromosome', i, 'is', length(SNP_chr), '\n')
      geno_chr = geno[ , SNP_chr]
      # geno_chr_others = geno[, SNP_chr_others]
      print(dim(geno_chr))
      SNPchrL[[i]] = geno_chr
      save(geno_chr, file = paste0("../Geno/SNPchrs/SNPchr", i, ".RData"))

  }
  # SNPchrLL[[j]] <- SNPchrL
  
# }


save(paste0("SNPchrL[[",1:12,"]]"), file = "../Geno/SNPchrs.RData")
```

### To check if the list file saved correctly.
```{r}
load("../Geno/SNPchrL.RData")
names(SNPchrL)
```